### PR TITLE
staticanalysis/bot: Limit clang-format patch generation to allowed paths.

### PR DESCRIPTION
**Pull request by La0**

### CODE REVIEW

**Comment by marco-c** [comment]
What if we only add contents to the patch from paths for which we have a issue which validates?

**Comment by La0** [comment]
@marco-c I'd like to still run the full `./mach clang-format`, and then only build a diff out of allowed issues. But i would need to parse the full patch & keep only allowed chunks (no time for that until tommorow's release :cry: )

**Comment by jankeromnes** [comment]
> What if we only add contents to the patch from paths for which we have a issue which validates?

That's a great suggestion! It could work for issue-generated improvement diffs, but we've changed `clang-format`'s improvement diff generation to use a single full patch directly from `./mach clang-format`.

The reason is that clang-format issues are inferred from diff hunks, but sometimes diff hunks don't map 1 to 1 with actual issues (e.g. in mass re-indents, often a random line with a `}` is picked as "unchanged", with everything above and below changing wildly), which produces subtle bugs and strange replacements like "replace this whole code by something totally unrelated".

Long story short, some `clang-format` issues/comments might still have subtle bugs occasionally, but now the improvement diff is guaranteed to be a real working patch. And filtering the patch on issue validity might be a challenge because they're no longer linked.

**Comment by marco-c** [comment]
If doing it for hunks is difficult, we can do it at the level of paths (similarly to this patch, but getting the paths from the ClangFormatIssues that validate).

**Comment by jankeromnes** [comment]
Interesting. But say you have 3 ClangFormatIssues for path `dom/media/CubebUtils.cpp`, and only 2 of them validate, what do you do? (I agree that's not possible right now for ClangFormatIssues, but it might happen for other types of issues, or if we change the validation algorithm in the future).

**Comment by marco-c** [comment]
> Interesting. But say you have 3 ClangFormatIssues for path dom/media/CubebUtils.cpp, and only 2 of them validate, what do you do? (I agree that's not possible right now for ClangFormatIssues, but it might happen for other types of issues, or if we change the validation algorithm in the future).

This problem also applies to the current situation, doesn't it? It wouldn't be changed by what I proposed, which is just a different way to select allowed paths.
Instead of applying the allowed paths filter everywhere, we'd just apply it in one place (`validate`). It only solves the problem of duplication of logic, but it doesn't solve the problem that some hunks should not be in the patch.

In the situation from your example, we'd show the full patch for dom/media/CubebUtils.cpp both with the implementation from this PR and with the implementation with `validates`.
